### PR TITLE
[OTBN, rtl] Introduce Blanking

### PR DESCRIPTION
**Pull request by GregAC**

### CODE REVIEW

**Comment by GregAC** [comment]
@vogelpi This implements the actual blanking and begins the implementation of pre-decode errors. This is where the enables used for the blanking generated by the pre-decode stage are checked against the outputs from the decode stage to ensure they're consistent.

**Comment by GregAC** [comment]
The RTL on this is now complete, though I should use `prim_buf` to prevent synthesis optimizing away blanking.

Leaving it as a draft as I should do some more tidy up and commenting before it's ready for merge. Along with some documentation.

File: hw/ip/otbn/rtl/otbn_decoder.sv
Comment (GregAC) [reply hw/ip/otbn/rtl/otbn_decoder.sv:521]: Yes, `BN.NOT` reads from the b register not the a. Before blanking was implemented these enables were pretty much unused (having been added in the first place in anticipation of blanking) so any errors in them went unnoticed.

File: hw/ip/otbn/rtl/otbn_instruction_fetch.sv
Comment (GregAC) [reply hw/ip/otbn/rtl/otbn_instruction_fetch.sv:112]: They're to do with the instructions that use indirect register addressing. Where we read from the base RF to get the register index for the bignum RF. I'll add some comments.

File: hw/ip/otbn/rtl/otbn_alu_bignum.sv
Comment (GregAC) [reply hw/ip/otbn/rtl/otbn_alu_bignum.sv:296]: I wanted to ensure the AND was directly applied to `operand_a`, though of course optimisation may move it around anyway. We'll need to use `prim_and2` here.

File: hw/ip/prim/rtl/prim_onehot_mux.sv
Comment (msfschaffner) [reply hw/ip/prim/rtl/prim_onehot_mux.sv:23]: If the buffer cells can be avoided by using a `prim_and2` directly, I would do that, yes.

I believe we have just recently added that primitive (both for generic and ASIC targets):
https://github.com/lowRISC/opentitan/blob/master/hw/ip/prim_generic/rtl/prim_generic_and2.sv

**Comment by GregAC** [comment]
@vogelpi @andreaskurth I've rebased my work on (almost) latest master as well as found and fixed a few bugs. Only failure coming up in the regression now is some kind of race condition between reading an invalid key from the key WSRs and a failure due to invalid ECC from the instruction memory. The DV environment thinks the bad instruction read should result in OTBN locking up whilst the RTL executes an instruction that reads the invalid key first so doesn't lock up just reports a KEY_INVALID error. It's possible this is caused by some timing differences introduced by blanking (though the prefetch stage is already in so it's only MOVR/SID that have adjusted timings) or just an existing DV/RTL bug that hasn't been fixed yet (@rswarbrick @prajwalaputtappa any thoughts?)

What still needs doing is the secure wipe work, doubly so because secure wipe is now enabled in master so this cannot be merged until that's implemented. It should be pretty straight-forward. `otbn_start_stop_control` will need a port into `otbn_instruction_fetch` so it can get at the pre-decoded register/ISPR write enables plus appropriate plumbing to ensure the zero/urnd bits are fed to the bignum register file at the appropriate cycle (everything will be one cycle after it was previously).

> We've decided not to blank the base RF/pipeline which I think is reasonable. But looking at the [block diagram](https://docs.opentitan.org/hw/ip/otbn/doc/index.html#block-diagram) I can see several signals running between the bignum ALU / MAC and the base pipeline. This means we probably need to apply blanking also to some bignum ALU / MAC outputs even though we say we don't because we do it on the input.

Those datapaths are from ISPRs not the output of the functional units. Those ISPRs already have blanking in their read muxes.

> You mentioned in the issue that we can't apply this to the FPGA regfile. I was wondering if we could apply it in a slightly different form, i.e., instead of gating the WDR outputs, apply the address input from the registers. This way we would at least not have glitches on the read addresses.

Good idea, well worth investigating, though not a priority right now. Let's open an issue for it so it's not forgotten.

File: hw/ip/otbn/rtl/otbn_controller.sv
Comment (GregAC) [inline hw/ip/otbn/rtl/otbn_controller.sv:1127]: I'm not sure if we need any blanking here as `lsu_addr_o` comes from the base side. You'd need blanking here if you'd worry about leaking information about addresses read and written.

**Comment by GregAC** [comment]
Oh and note first two commits come from https://github.com/lowRISC/opentitan/pull/11939

File: hw/ip/otbn/rtl/otbn_core.sv
Comment (GregAC) [inline hw/ip/otbn/rtl/otbn_core.sv:352]: The logic for this is no complex but I only just remember I wanted to add it so haven't yet tested the no read enables if there's no valid instruction holds. It is possible there will be some bugs to fix once this is added.

**Comment by GregAC** [comment]
#11939 has now been merged I've rebased this PR to remove those commits.

File: hw/ip/otbn/rtl/otbn_mac_bignum.sv
Comment (GregAC) [inline hw/ip/otbn/rtl/otbn_mac_bignum.sv:117]: This needs a prim_blanker

File: hw/ip/otbn/rtl/otbn_alu_bignum.sv
Comment (vogelpi) [inline hw/ip/otbn/rtl/otbn_alu_bignum.sv:154]: Nit: could align the equal operator with the previous statement on Line 150.

File: hw/ip/otbn/rtl/otbn_alu_bignum.sv
Comment (vogelpi) [inline hw/ip/otbn/rtl/otbn_alu_bignum.sv:434]: Nit: I would maybe abbreviate `expected` here by `exp` as these signal names are pretty long. Doing that now when the signals are added is certainly easier than later.

File: hw/ip/otbn/rtl/otbn_alu_bignum.sv
Comment (vogelpi) [inline hw/ip/otbn/rtl/otbn_alu_bignum.sv:395]: Is this blanker really needed? Can't you just re-use `u_adder_x_op_a_blanked`? At least your comment on Line 40 says that Adder X is always used with Adder Y. Maybe I just misunderstand something here.

File: hw/ip/otbn/rtl/otbn_alu_bignum.sv
Comment (vogelpi) [inline hw/ip/otbn/rtl/otbn_alu_bignum.sv:333]: Would it make sense - or is it even possible - to combine this mux here with the blanker above? Because if `shifter_b_en` is 0, the blanker will also output 0.

File: hw/ip/otbn/rtl/otbn_alu_bignum.sv
Comment (vogelpi) [inline hw/ip/otbn/rtl/otbn_alu_bignum.sv:274]: Nit: Somehow I find the previous code with the line break after the "zeros" more readable, but this is probably a personal preference.

File: hw/ip/otbn/rtl/otbn_alu_bignum.sv
Comment (vogelpi) [inline hw/ip/otbn/rtl/otbn_alu_bignum.sv:249]: Is this still something that needs to be done?

File: hw/ip/otbn/rtl/otbn_alu_bignum.sv
Comment (vogelpi) [inline hw/ip/otbn/rtl/otbn_alu_bignum.sv:246]: Nit: I believe indentation should be 4 chars here.

**Review by vogelpi — COMMENTED** [review]
I've reviewed the latest version of this PR (except for otbn_core.sv and otbn_controller.sv which follows asap). This looks good. So far, I have only a couple of questions and nits.

File: hw/ip/otbn/rtl/otbn_controller.sv
Comment (vogelpi) [reply hw/ip/otbn/rtl/otbn_controller.sv:1135]: I think the address itself is not critical but what happens if the address is propagated to the memory. It will depend on the internals of the memory macro whether the look-up only happens with the `req` enabled or not. To be on the safe side, we should probably blank the address.

Side note: I was just thinking that maybe it would be sufficient to just blank the `lsu_addr_sel_o` signal such that by default the previous address from `lsu_addr_q` was used. But this wouldn't work as in case `alu_base_operation_result_i[DmemAddrWidth-1:0]` is selected we would still forward potential glitches on it as it's not blanked.

File: hw/ip/otbn/rtl/otbn_controller.sv
Comment (vogelpi) [inline hw/ip/otbn/rtl/otbn_controller.sv:1151]: I think the `_q` suffix here may be misleading as the signal is not flopped.

File: hw/ip/otbn/rtl/otbn_core.sv
Comment (vogelpi) [inline hw/ip/otbn/rtl/otbn_core.sv:353]: Is this signal correctly generated? Don't you need to compare the signals from the predecoder with the ones from the actual decoder to decide whether there was an error or not?

**Review by vogelpi — APPROVED** [review]
I've now also reviewed the remaining code parts. In general, this looks pretty good modulo some minor things and questions. Since I will be OoO tomorrow and I don't want to keep blocking this, I am approving the PR as is assuming my feedback will be addressed before merging.

File: hw/ip/otbn/rtl/otbn_predecode.sv
Comment (GregAC) [reply hw/ip/otbn/rtl/otbn_predecode.sv:184]: Believe they should have been removed. These enables will be set in the second cycle of the instruction due to the indirect register access so predecode doesn't need to set them here.

File: hw/ip/otbn/doc/_index.md
Comment (GregAC) [reply hw/ip/otbn/doc/_index.md:521]: It would use less logic but then when you write your new write data will be presented to the write port of every register in the register file. Perhaps this has the potential to cause extra leakage. I believe this is what they did in the Coco paper (https://eprint.iacr.org/2020/1294.pdf) where they applied similar measures to Ibex.

File: hw/ip/otbn/rtl/otbn_alu_bignum.sv
Comment (GregAC) [reply hw/ip/otbn/rtl/otbn_alu_bignum.sv:434]: My preference is to avoid abbreviations where possible unless they're very common (e.g. en for enable). I don't feel the names are excessively long here, in particular we're not needing line-breaks to stay within the line limit in expressions that are using them.

File: hw/ip/otbn/rtl/otbn_alu_bignum.sv
Comment (GregAC) [reply hw/ip/otbn/rtl/otbn_alu_bignum.sv:395]: It certainly touches on the overkill side but maybe not entirely pointless. When Adder X and Adder Y are used together the adder Y input is from Adder X, the operand A input isn't used by Adder Y. We only use the A operand directly with Adder Y when we use Adder Y alone.

When we're using Adder X and Adder Y operand A generating a power signature it doesn't need to in the mux above Adder Y (see text diagram at top of this file) may not be a problem as it will generated a power signature in Adder X, which it has to do.

Again when we're using Adder Y only A will generate a power signature in adder X (which it doesn't need to do) and a power signature in the mux above Adder Y (which it does need to do). Perhaps also not a problem?

File: hw/ip/otbn/rtl/otbn_alu_bignum.sv
Comment (GregAC) [reply hw/ip/otbn/rtl/otbn_alu_bignum.sv:246]: Ah yes probably. I've just decided to reformat it anyway to one term per line so it matches the expression above.

File: hw/ip/otbn/rtl/otbn_controller.sv
Comment (GregAC) [reply hw/ip/otbn/rtl/otbn_controller.sv:917]: Yup, have switched it to a simple ternary

File: hw/ip/otbn/rtl/otbn_core.sv
Comment (GregAC) [reply hw/ip/otbn/rtl/otbn_core.sv:353]: Yes it's specifically an error triggered when we attempt to read from the register file without a valid instruction. We can't compare with the actual decoder as without a valid instruction it's not producing valid bits.

We may want to add some checks around the write enable though. This will be set when we don't have a valid instruction (for secure wipe). We're doing something similar on the Ibex side.

File: hw/ip/otbn/rtl/otbn_controller.sv
Comment (GregAC) [reply hw/ip/otbn/rtl/otbn_controller.sv:1151]: Good spot

File: hw/ip/otbn/rtl/otbn_alu_bignum.sv
Comment (GregAC) [reply hw/ip/otbn/rtl/otbn_alu_bignum.sv:274]: I'd agree actually (despite writing this), I will change it.

File: hw/ip/otbn/rtl/otbn_controller.sv
Comment (GregAC) [reply hw/ip/otbn/rtl/otbn_controller.sv]: I've added blanking to this signal

File: hw/ip/otbn/rtl/otbn_alu_bignum.sv
Comment (GregAC) [reply hw/ip/otbn/rtl/otbn_alu_bignum.sv:334]: Just realised I forgot to address this, yes I think this can be done.

**Comment by GregAC** [comment]
I'm merging this now.

The points @vogelpi raises around blanking write data into the register file and the blanking of operand A into adder Y in the ALU should continue to be discussed. We can make any adjustments in a follow up PR.

File: hw/ip/otbn/rtl/otbn_alu_bignum.sv
Comment (vogelpi) [reply hw/ip/otbn/rtl/otbn_alu_bignum.sv:446]: Sounds good to me Greg.

File: hw/ip/otbn/rtl/otbn_core.sv
Comment (vogelpi) [reply hw/ip/otbn/rtl/otbn_core.sv]: Ah, now this makes sense! Thanks for explaining and adding comments in the RTL.

**Pull request merged**
